src/main/java/kafka/vertx/demo/WebSocketServer.java

```diff
@@ -0,0 +1,125 @@
+import io.vertx.config.ConfigRetrieverOptions;
+import io.vertx.config.ConfigStoreOptions;
+import io.vertx.core.AbstractVerticle;
+import io.vertx.core.Promise;
+import io.vertx.core.eventbus.EventBus;
+import io.vertx.core.http.ServerWebSocket;
+import io.vertx.core.json.JsonObject;
+import io.vertx.ext.web.Router;
+import io.vertx.ext.web.handler.StaticHandler;
+import io.vertx.kafka.client.common.TopicPartition;
+import io.vertx.kafka.client.consumer.KafkaConsumer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+
+public class WebSocketServer extends AbstractVerticle {
+
+  private static final String PRODUCE_PATH = "/demoproduce";
+  private static final String CONSUME_PATH = "/democonsume";
+
+  private static final Logger logger = LoggerFactory.getLogger(WebSocketServer.class);
+
+  private HashMap<String, String> kafkaConfig;
+
+  @Override
+  public void start(Promise<Void> startPromise) {
+    loadKafkaConfig(startPromise);
+
+    Router router = Router.router(vertx);
+    router.get().handler(StaticHandler.create());
+
+    vertx.createHttpServer()
+      .requestHandler(router)
+      .webSocketHandler(this::handleWebSocket)
+      .listen(8080, "localhost")
+      .onSuccess(ok -> {
+        logger.info("🚀 WebSocketServer started");
+        startPromise.complete();
+      })
+      .onFailure(err -> {
+        logger.error("❌ WebSocketServer failed to listen", err);
+        startPromise.fail(err);
+      });
+  }
+
+  private void loadKafkaConfig(Promise<Void> startPromise) {
+    ConfigRetriever.create(vertx,
+      new ConfigRetrieverOptions().addStore(
+        new ConfigStoreOptions()
+          .setType("file")
+          .setFormat("properties")
+          .setConfig(new JsonObject().put("path", "kafka.properties").put("raw-data", true))))
+      .getConfig()
+      .onSuccess(config -> {
+        kafkaConfig = new HashMap<>();
+        config.forEach(entry -> kafkaConfig.put(entry.getKey(), entry.getValue().toString()));
+      })
+      .onFailure(startPromise::fail);
+  }
+
+  private void handleWebSocket(ServerWebSocket webSocket) {
+    switch (webSocket.path()) {
+      case PRODUCE_PATH:
+        handleProduceSocket(webSocket);
+        break;
+      case CONSUME_PATH:
+        handleConsumeSocket(webSocket);
+        break;
+      default:
+        webSocket.reject();
+    }
+  }
+
+  private void handleProduceSocket(ServerWebSocket webSocket) {
+    EventBus eventBus = vertx.eventBus();
+
+    webSocket.handler(buffer -> {
+      String action = buffer.toJsonObject().getString("action", "none");
+      if ("start".equals(action) || "stop".equals(action)) {
+        eventBus.send(Main.PERIODIC_PRODUCER_ADDRESS, action);
+      }
+    });
+
+    eventBus.<JsonObject>consumer(Main.PERIODIC_PRODUCER_BROADCAST, message ->
+      eventBus.send(webSocket.textHandlerID(), message.body().encode()));
+  }
+
+  private void handleConsumeSocket(ServerWebSocket webSocket) {
```

Comment:
I can see that this does do the equivalent of what we had before. A bug I recently found but haven't gotten around to addressing is how to stop the producer and consumer when the web socket disconnects, so the app doesn't continue to consume under the covers. Do you think it would be easier to add that behaviour with the consumer code still in a separate verticle? Or will the consumer get removed when the websocket is disconnected, since it is instantiated inside the handler?

Reply:
Good point, we should use `endHandler` on the web socket to close the consumer, as it will not magically disappear 👍